plugins/modules/source_control/gitlab/gitlab_service.py

```diff
@@ -0,0 +1,516 @@
+        diff = not self.__equals(remote.attributes, active, params, events)
+        if diff:
+            # remote.active = active
+            for k, v in self.__expand_events(events).items():
+                setattr(remote, k, v)
+            # when saving, params does directly in the request
+            # (remember they get retrieved under the "property" key of the remote dict())
+            for k, v in params.items():
+                setattr(remote, k, v)
+            if not self._module.check_mode:
+                remote.save()
+            return remote
+
+        return False
+
+    def __as_api_object(self, active, params, events):
+        obj = {'active': active}
+        obj.update(self.__expand_events(events))
+        obj.update(params)
+        return obj
+
+    def __expand_events(self, events):
+        try:
+            # See https://gitlab.com/gitlab-org/gitlab-ce/issues/58321 for why it's useful to
+            # discard unsupported events before comparing
+            supported_events = SRV_DEF[self.name]['_events']
+        except KeyError:
+            supported_events = None
+        ret = dict((value + '_events', bool(value in events)) for value in self.HOOK_EVENTS if supported_events is None or value in supported_events)
+        return ret
+
+    @staticmethod
+    def __events_equal(obj, events):
+        return all((k in obj and obj[k] == v) for k, v in events.items())
+
+    @staticmethod
+    def __has_credentials(attr):
+        return ('password' in attr and attr['password']) or ('token' in attr and attr['token']) or ('api_key' in attr and attr['api_key'])
+
+    def __equals(self, prev_attr, active, params, events):
+        filtered_params = dict((k, v) for k, v in params.items() if k not in self.CREDENTIAL_PARAMS)
+        if prev_attr['active'] != active:
+            self._module.debug("active status differs")
+            return False
+        if self.__has_credentials(params):
+            self._module.debug("has credentials: assuming difference")
+            return False
+        if not self.__events_equal(prev_attr, self.__expand_events(events)):
+            self._module.debug("events differs: %s != %s" % (prev_attr, events))
+            return False
+        if prev_attr['properties'] != filtered_params:
+            self._module.debug("services attributes differs: %s != %s" % (prev_attr['properties'], filtered_params))
+            return False
+        return True
+
+    def setattrs(self, obj, new_attr):
+        for k, v in new_attr.items():
+            setattr(obj, k, v)
+
+
+def main():
+    definitions = init_definitions()
+
+    base_spec = basic_auth_argument_spec()
+    base_spec.update(dict(
+        api_token=dict(type='str', no_log=True),
+        project=dict(required=True),
+        service=dict(required=True, type='str', choices=list(definitions.keys())),
+        # active=dict(required=False, default=True, type='bool'),
+        params=dict(required=False, type='dict'),
```

Comment:
```suggestion
        params=dict(required=False, type='dict', no_log=True),
```
Since this can contain secrets, it **must** be marked `no_log=True`.